ls コマンドを作るオブジェクト指向版のプログラムを作成

a-kuroki-gs opened the pull request:
フィヨルドブートキャンプの[ls コマンドを作るオブジェクト指向版](https://bootcamp.fjord.jp/practices/213)の課題です。

a-kuroki-gs committed:
lオプションを実装

a-kuroki-gs committed:
lオプションなしを実装

a-kuroki-gs committed:
rubocop対応

a-kuroki-gs committed:
Fileクラスでは変更時間をフォーマット化しない

a-kuroki-gs committed:
コメントの削除

a-kuroki-gs committed:
クラス内で使用するメソッドをprivateにする

JunichiIto commented on 07.ls_object/run_ls.rb:
```diff
@@ -0,0 +1,27 @@
+opt.parse!(ARGV)
+directory = ARGV[0] || '.'
+
+d = Directory.new(directory)
```
寿命の長い変数は略さないようにした方が良いです
https://blog.jnito.com/entry/2020/10/20/092724

a-kuroki-gs replied:
変更しました(が、その後使わなくなりました)

JunichiIto commented on 07.ls_object/run_ls.rb:
```diff
@@ -0,0 +1,27 @@
+
+d = Directory.new(directory)
+
+d = d.reject_dot_files unless params[:a]
```
```suggestion
d.reject_dot_files unless params[:a]
```

再代入しなくても`reject_dot_files`を呼べばインスタンスの状態（`@files`内のファイル）が変わるようになっていますね。
ただし、「状態がコロコロ変わっても自然なクラス」でなければ、なるべくインスタンスの状態は変えない方が良いです。

「状態がコロコロ変わっても自然なクラス」というのはたとえば「信号機クラス」などが考えられます。
信号機であれば赤、黄、緑という3つの状態があって、それが随時変わるというのは考え方として自然です。

一方、ディレクトリは実際にファイルを追加したり削除したりしないと状態（ファイルの件数）は変わらないですし、ディレクトリ自体が固定の並び順を持っているわけではない（ターミナル上で一度ソートしたらその後もずっとその状態が維持される、なんてことはない）ので、ディレクトリクラス内の`@files`が増えたり減ったり、並び順が変わったりするのは、「ディレクトリ」というオブジェクトを考えたときにあまり自然ではありません。

`reject_dot_files`や`reverse_files`はディレクトリというより、lsコマンドとして必要なメソッドだと考えた方が良いです。
（つまりディレクトリクラスの責務ではない）

a-kuroki-gs replied:
再代入の意味があまりない気がする…と思っていた部分でした。
ご指摘ありがとうございます。
ここは`display_files`として、ファイルの中身ではなく表示する内容を格納するようにしました。

JunichiIto commented on 07.ls_object/run_ls.rb:
```diff
@@ -0,0 +1,27 @@
+
+if params[:l]
+  puts "合計 #{d.calculate_block_counts}"
+  puts d.print_l_option
```
今の設計だと、ディレクトリクラスには`@files`というデータを管理する責務と、ターミナルにファイルのデータを表示するという2つの責務を持っています。
データと画面表示はなるべく分けた方が再利用性が高くなるので、同じクラスに含めない方がいいです。

たとえば、ターミナルではなくブラウザに表示したくなったら、今の設計だとDirectoryクラスにHTML出力用のメソッドが増えることになり、どんどんDirectoryクラスがfatになってしまいます。

a-kuroki-gs replied:
ファイルのデータを管理する`Directory`クラスと、データを表示する`Display`クラスに分割しました。

JunichiIto commented on 07.ls_object/file.rb:
```diff
@@ -0,0 +1,74 @@
+class File
+  def initialize(path)
+    @path = path
+    @file = File.lstat(@path)
```
lstatの戻り値をfileに代入すると、コードの読み手が混乱しそうです。
https://bootcamp.fjord.jp/pages/readable-variable-name

a-kuroki-gs replied:
`file_stat`に変更しました

JunichiIto commented on 07.ls_object/file.rb:
```diff
@@ -0,0 +1,74 @@
+  def initialize(path)
+    @path = path
+    @file = File.lstat(@path)
+    @file_mode = @file.mode.to_s(8)
```
この変数はmodeメソッドを呼び出したときにしか使わないので、インスタンス変数に入れなくてもいい気がしました。

a-kuroki-gs replied:
ここを消去し、必要なときに直接呼び出すように修正しました。

JunichiIto commented on 07.ls_object/directory.rb:
```diff
@@ -0,0 +1,71 @@
+
+  def initialize(directory)
+    @files =
+      Dir.entries(directory).sort.each_with_object([]) do |file, files|
```
mapメソッドを使った方がシンプルになると思いますー。

a-kuroki-gs replied:
`map`メソッドを利用し書き換えました。

JunichiIto commented on 07.ls_object/directory.rb:
```diff
@@ -0,0 +1,71 @@
+        " #{file.mtime.strftime('%-m月').rjust(5)}",
+        " #{file.mtime.strftime('%e %H:%M')}",
+        " #{file.name}"
+      ].join.rstrip
```
コードを見る限り、rstripが必要になるケースはないように思います。

a-kuroki-gs replied:
必要なかったので削除しました。

JunichiIto commented on 07.ls_object/directory.rb:
```diff
@@ -0,0 +1,71 @@
+    @files.sum(&:blocks) / 2
+  end
+
+  def print_not_l_option
```
not l optionという名前がちょっと苦しい感じがしますね。
たとえば「左に曲がれ」というのを"turn not right"とは言ったりしないので、できればnotを使わない名前にしたいです。

ターミナルで`man ls`を実行してlオプションの説明を読んでみてください。
その説明を参考にして、まず`print_l_option`について、もっと良い名前を考えてみてください。

それからこの`print_not_l_option`に戻ってくると、より適切な名前が思いつくかもしれません。

a-kuroki-gs replied:
`print_l_option`を`print_detailed_list_format`に、
`print_not_l_option`を`print_simple_list_format`に変更しました。

JunichiIto reviewed (commented):
コメントしました〜。

a-kuroki-gs committed:
読みにくいコードを修正

寿命の長い変数は略さないようにする
機能していないrstripを消去
変数化するほどでないものは変数にしない
ごちゃごちゃしていた配列とクラスの処理にmapメソッドを使う
File::Statのオブジェクトに対し、file_statの変数名を使う

a-kuroki-gs committed:
わかりにくいメソッド名の修正

表示のところでnot_l_optionなどとするのではなく、print_detailed_list_format・print_simple_list_formatとする

a-kuroki-gs committed:
Directoryクラスが機能を持ちすぎていたので分ける

ファイルのデータを管理するDirectoryクラスと、データを表示するDisplayクラスに分ける

JunichiIto commented on 07.ls_object/directory.rb:
```diff
@@ -0,0 +1,15 @@
+      Dir.entries(directory).sort.map do |file|
+        path = "#{directory}/#{file}"
+        File.new(path)
+      end
```
結局のところ、DirectoryクラスはFile情報を配列に詰め込むだけの仕事しかしてないので、「じゃあそれはそのまま配列として使えばええやん」という話になります。（Directoryとして他に何か特別なメソッドを持っているなら別ですが）
なので、Directoryクラスは無しにしてしまいましょう。

a-kuroki-gs replied:
Directoryクラスを削除し、取得したファイルは配列に格納しました。

JunichiIto commented on 07.ls_object/file.rb:
```diff
@@ -0,0 +1,73 @@
+
+require 'etc'
+
+class File
```
前回コメントし忘れていましたが、RubyにはすでにFileクラスが存在しています。
https://docs.ruby-lang.org/ja/latest/class/File.html

なので、今のコードだと、Ruby標準のFileクラスをkurokiさんが独自に拡張している状態になっています。

おそらくそれは意図してやってないと思うので、Ruby標準のクラスと被らない名前を付けてください。

a-kuroki-gs replied:
被らないように`FileStat`クラスにしました。
(`File::Stat`と似ていますが)

JunichiIto commented on 07.ls_object/run_ls.rb:
```diff
@@ -0,0 +1,26 @@
+opt.parse!(ARGV)
+input = ARGV[0] || '.'
+
+display_files = Display.new(input)
```
```suggestion
display = Display.new(input)
```

左辺と右辺の名前は原則揃えましょう。
https://bootcamp.fjord.jp/pages/readable-variable-name

あと、inputという変数名は無味無臭なので、中身が想像できる変数名の方がいいですね。
https://twitter.com/jnchito/status/1571287095574138881

a-kuroki-gs replied:
それぞれ変数名の変更を行いました。

JunichiIto commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,65 @@
+
+  COLUMN_NUMBER = 3
+
+  def reject_dot_files
```
クラスを設計するときはそのクラスの責務を考えることが大事です。
Displayというクラス名から想像できる責務はデータの表示かなと思います。
なので、`print_simple_list_format`や`print_detailed_list_format`が定義されているのは自然ですが、それ以外のメソッドはDisplayクラスの責務とはちょっと違うんじゃないかなーと思いました。

a-kuroki-gs replied:
クラス名と中身の差異があったため
表示するクラスと、その前段階の処理のクラスに分けました。

JunichiIto commented on 07.ls_object/run_ls.rb:
```diff
@@ -0,0 +1,26 @@
+display_files = display_files.reverse_files if params[:r]
+
+if params[:l]
+  puts "合計 #{display_files.calculate_block_counts}"
+  puts display_files.print_detailed_list_format
```
```suggestion
  puts display_files.print_detailed_list_format
```

合計の出力も `print_detailed_list_format` の中でやった方がシンプルになりそうです

a-kuroki-gs replied:
表示もメソッド内で完結させるよう修正しました。

JunichiIto reviewed (commented):
コメントしました〜。

JunichiIto commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,65 @@
+    @display_files.sum(&:blocks) / 2
+  end
+
+  def print_simple_list_format
```
`print_simple_list`、`print_detailed_list`ぐらいの方が短くて良い気がしました。

ただ、このメソッドって、べつにprintしてくれるわけではないんですよね？そこはちょっと違和感がありますね 🤔

a-kuroki-gs replied:
メソッド名と合わせて表示を行うよう修正しました。
(また、`print`をクラス名と合わせて`display`に変更しております)

a-kuroki-gs committed:
表示するメソッドの修正

メソッド名と内容を合わせるため

a-kuroki-gs committed:
変数名の変更

左辺と右辺をそろえるため
inputが無味無臭であるため

a-kuroki-gs committed:
Fileクラスの名前を変更

RubyにはすでにFileクラスが存在しているため

a-kuroki-gs committed:
結果を表示するクラスと、その前段階の操作のクラスに分ける

クラス名と中身の差異があったため

a-kuroki-gs committed:
rubocop対応

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

a-kuroki-gs committed:
定数をクラスの頭にもってくる

a-kuroki-gs committed:
表示するメソッドの先頭をprintに変更

クラス名やインスタンス名と被らないようにするため

a-kuroki-gs committed:
変数名の修正

インスタンス名はクラス名に合わせてつけるようにする
その他分かりにくいものも修正

a-kuroki-gs committed:
mapの使い方が間違っていたので修正

a-kuroki-gs committed:
不要な改行の削除

a-kuroki-gs committed:
joinの引数にスペースを与えて、すっきりとまとめる

a-kuroki-gs committed:
FileManagerクラスを削除する

あまり難しいことは行っていなかったため
a・rオプションは配列に対してrun_ls.rbで直接実行するようにする
lオプションのブロック数はDisplayクラス内で計算する

a-kuroki-gs committed:
変数名の見直し

JunichiIto commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,49 @@
+    end
+
+    before_format_display = @file_stats.map(&:name).each_slice(row_number).to_a
+    before_format_display.last.size == row_number || before_format_display.last.fill(nil, before_format_display.last.size...row_number)
```
何で`||`を使ってるんだろう？と思ったんですが、これは条件分岐なんですね。
`&&`は`||`は真偽値を返すときに使うことが多いので、`||`を使わない書き方に修正してみてください〜。

a-kuroki-gs replied:
`unless`を使って書き換えました。

JunichiIto commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,49 @@
+        file.user.ljust(max_user),
+        file.group.ljust(max_group),
+        file.bytesize.to_s.rjust(max_size),
+        file.mtime.strftime('%-m月').rjust(5),
+        file.mtime.strftime('%e %H:%M'),
```
1行にまとめることはできないですかね〜？

```suggestion
        file.mtime.strftime(????),
```

a-kuroki-gs replied:
こちら１桁の月と２桁の月をきれいに揃えるのが難しいと判断し、操作を分けておりましたが、１行で記載することも可能です。
方法として、この２行を連結する、もしくは`format`(`sprintf`)で揃えてあげる、という２種類があるかと思います。
今回は`format`を利用して１行で出力するようにしてみましたが、この修正でよろしいでしょうか？

JunichiIto commented on 07.ls_object/run_ls.rb:
```diff
@@ -0,0 +1,31 @@
+  FileStat.new(path)
+end
+
+file_stats = file_stats.reject { |file_stat| file_stat.name.start_with?('.') } unless params[:a]
+file_stats = file_stats.reverse if params[:r]
+
+display = Display.new(file_stats)
+
+if params[:l]
+  display.print_detailed_list
+else
+  display.print_simple_list
+end
```
このへんの処理、Displayクラスに入れちゃってもいいかも

```suggestion
display = Display.new(file_stats, params)
display.print_list
```

a-kuroki-gs replied:
`Display`クラスのインスタンス作成時に引数として`params`を渡し、
オプションによる操作を`Display`クラス内で行うように修正しました。

JunichiIto reviewed (commented):
コメントしました〜。

a-kuroki-gs committed:
日時を表示しているところを１行で記載する

a-kuroki-gs committed:
オプションを受け取って処理を行う部分をDisplayクラスに入れ込んで行う

a-kuroki-gs committed:
||で条件分岐をしていたところをunlessを利用するよう修正

a-kuroki-gs committed:
変数名を修正

a-kuroki-gs committed:
rubocopを通す

JunichiIto commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,60 @@
+    @params = params
+  end
+
+  def print_simple_list
```
print_list が導入されたので、privateメソッドにした方がいいですね

a-kuroki-gs replied:
`print_simple_list`と`print_detailed_list`をprivateメソッドに修正しました。

JunichiIto commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,60 @@
+    end
+
+    before_format_display = @file_stats.map(&:name).each_slice(row_number).to_a
+    before_format_display.last.fill(nil, before_format_display.last.size...row_number) unless before_format_display.last.size == row_number
```
```suggestion
    last_col = before_format_display.last
    last_col.fill(nil, last_col.size...row_number) unless last_col.size == row_number
```

みたいに変数に入れた方が見やすくなりそうです

a-kuroki-gs replied:
ご指摘いただいたよう修正しました。

JunichiIto commented on 07.ls_object/display.rb:
```diff
@@ -0,0 +1,60 @@
+      file_stat.map { |file| file.name.size }.max
+    end
+
+    before_format_display = @file_stats.map(&:name).each_slice(row_number).to_a
```
displayという名前を付けるとこのクラス（Displayクラス）のインスタンスが含まれているように見えます。
each_sliceを呼ぶことでファイル名の配列が入れ子になる（ネストする）ので、 `nested_file_names`みたいな名前がいいかもしれません

a-kuroki-gs replied:
`display`がいくつか残っており、見にくいコードになっておりました。

`before_format_display`→`nested_file_names`
`formatted_display`→`formatted_file_names`
(`each`の中の)`display`→`file_names`
のように修正しました。

JunichiIto commented on 07.ls_object/file_stat.rb:
```diff
@@ -0,0 +1,73 @@
+  end
+
+  def group
+    Etc.getpwuid(@lstat.gid).name
```
僕のローカルで実行するとエラーになりました。原因を考えてみてください。
（kurokiさんの環境では偶然運良く動いてるだけっぽい）

```
$ ruby run_ls.rb -l
/Users/jnito/dev/fjord/a-kuroki-gs/ruby-practices/07.ls_object/file_stat.rb:45:in `getpwuid': can't find user for 20 (ArgumentError)

    Etc.getpwuid(@lstat.gid).name
                 ^^^^^^^^^^
	from /Users/jnito/dev/fjord/a-kuroki-gs/ruby-practices/07.ls_object/file_stat.rb:45:in `group'
	from /Users/jnito/dev/fjord/a-kuroki-gs/ruby-practices/07.ls_object/display.rb:33:in `block in print_detailed_list'
	from /Users/jnito/dev/fjord/a-kuroki-gs/ruby-practices/07.ls_object/display.rb:33:in `map'
	from /Users/jnito/dev/fjord/a-kuroki-gs/ruby-practices/07.ls_object/display.rb:33:in `print_detailed_list'
	from /Users/jnito/dev/fjord/a-kuroki-gs/ruby-practices/07.ls_object/display.rb:55:in `print_list'
	from run_ls.rb:23:in `<main>'
```

a-kuroki-gs replied:
`getpwuid`は`uid`からユーザを探してくれるメソッドであり、
`gid`からグループを探してくれるのは`getgrgid`メソッドであることがわかりました。

再現はできていませんが、`getgrgid`メソッドに修正しております。

JunichiIto commented on 07.ls_object/run_ls.rb:
```diff
@@ -0,0 +1,23 @@
+end
+
+display = Display.new(file_stats, params)
+display.print_list
```
```ruby
display.print_list
puts
display.print_list
```

みたいに `print_list` を2回呼ぶようにして、 r オプション付きで実行すると、1回目と2回目で並び順が変わります。
（2回目の表示がおかしい）
この問題が起きないようにコードを修正してください。

```
$ ruby run_ls.rb -lr
合計 12
-rw-r--r-- 1 jnito staff  544 12月 26 10:28 run_ls.rb
-rw-r--r-- 1 jnito staff  971 12月 26 10:28 file_stat.rb
-rw-r--r-- 1 jnito staff 1944 12月 26 10:25 display.rb

合計 12
-rw-r--r-- 1 jnito staff 1944 12月 26 10:25 display.rb
-rw-r--r-- 1 jnito staff  971 12月 26 10:28 file_stat.rb
-rw-r--r-- 1 jnito staff  544 12月 26 10:28 run_ls.rb
```

a-kuroki-gs replied:
`print_list `メソッド内で`-a`・`-r`オプションの処理を行っていたことが問題でした。

`-a`・`-r`オプションの処理は`initialize`メソッド内に移動しております。
またそれに伴い一部ロジックを変更し、無駄な代入を行わないよう修正しました。

JunichiIto reviewed (commented):
コメントしました〜。

a-kuroki-gs committed:
外部からアクセスする必要のないメソッドをprivateにする

a-kuroki-gs committed:
１文が長くなりすぎていたところを修正

変数を用いることで２文にわけて、無駄な処理を減らしわかりやすくする

a-kuroki-gs committed:
変数名の変更

Displayクラスのインスタンスでないところに、displayをつけないようにする

a-kuroki-gs committed:
a・rオプションの処理を表示するメソッドから、initializeメソッド内に移動

表示とは別の操作であるため
表示の中に入れてしまうと、仕様通り動くのは最初の１度しか保証されないため
また、一部ロジックの修正を行い、無駄な代入を削除した

a-kuroki-gs committed:
gidからグループを取り出すメソッドでgetpwuidを使ってしまっていたのでgetgrgidに修正

JunichiIto commented on 07.ls_object/display.rb:
```diff
@@ -4,24 +4,41 @@ class Display
+    nested_file_names = @file_stats.map(&:name).each_slice(row_number).to_a
+    last_col = nested_file_names.last
+    last_col.fill(nil, last_col.size...row_number) unless last_col.size == row_number
+    formatted_file_names = nested_file_names.transpose
```
formatを使うとファイル名の表示方法（たとえば右寄せ、左寄せとか）を整形してような印象がありますね。
命名が難しいところですが、ファイル名が縦と横に並んだ表＝テーブルと見なして、

```suggestion
    file_name_table = nested_file_names.transpose
```

みたいにするといいかもしれません

JunichiIto reviewed (approved):
1点コメントしましたが大きな問題ではないのでこれでOKです！🙆‍♂️
